### ISSUE TITLE
(params_with_wd if needs_wd else params_with_wd)['params'].append(parameter) in deep.py，is there a mistake here? Why if...else...  are all connected to params_with_wd?

### ISSUE THREAD

**Issue by luozhengdong** [yandex-research/tabular-dl-tabr#6]
Thanks for your interesting work, which gives us new hope for tabular data research!  
I have some questions  to ask:
（1）(params_with_wd if needs_wd else params_with_wd)['params'].append(parameter) in deep.py，is there a mistake here? Why if...else...  are all connected to params_with_wd?
（2）This bug occurs during training when trial_0 is completed and trial_1 is in progress, I observed that the gpu usage increases as trial increases, is this normal? Is it possible to release the gpu after trial_0 is completed and then proceed to trial_1? Currently, we only have 12 G of gpu resources.
[...] tmp22mqfk7j_trial_1/output | 0:03:04.846372
Epoch 39: 100%|███████████████████████████████████████████████████████████████████| 52/52 [00:04<00:00, 12.
(val) -0.426 (test) -0.414 (loss) 0.11597█████████████████████████████████████████| 52/52 [00:04<00:00, 13.
[W 2023-10-24 17:37:27,179] Trial 1 failed with parameters: {'model.d_main': 353, 'model.context_dropout': 76563006176, 'model.dropout0': 0.2300649112954666, 'optimizer.lr': 0.0004817508474772368, '?optimizer.weigh': True, 'optimizer.weight_decay': 7.098936257405907e-05} because of the following error: AttributeError("mtorch.cuda' has no attribute 'OutOfMemoryError'").

**Comment by Yura52** [comment]
Hi! Thank you for your interest in our work and for reporting these issues!

> is there a mistake here? Why if...else... are all connected to params_with_wd?

Yes, this is definitely a mistake :) Luckily, this is not something critical, though we should remove this to avoid confusion.

> I observed that the gpu usage increases as trial increases, is this normal?

Does it happen only between trial 0 and trial 1 or throughout the whole tuning process? Also, is your environment configured as described in README? I mean all the software versions etc. I have just checked my environment for TabR and `torch.cuda` has the `OutOfMemoryError` class in it, however, in a different environment, it does not. Perhaps, `torch.cuda.OutOfMemoryError` is available only in some versions of PyTorch and only if CUDA is available.

> Currently, we only have 12 G of gpu resources.

Most of the time I run experiments on a 12G GPU, too, and so far, without any issues related to memory leaks. Overall, for now, I don't have good explanations for why it can happen.

**Comment by luozhengdong** [comment]
What is the difference or connection between trial_0 and trial_1?The process of trial_0 consists of: each epoch contains batch number as samples number /batch size, the epochs number performed depends on the patience(set as 16) setting, then reprot and save.... 
So trial_1 is also the same process  as trail_0?

**Comment by Yura52** [comment]
Yes. Let's consider [this config](https://github.com/yandex-research/tabular-dl-tabr/blob/d628ec7e1c0a66011473021034e7dd4a77740112/exp/tabr/california/0-tuning.toml) as an example:

- [this line](https://github.com/yandex-research/tabular-dl-tabr/blob/d628ec7e1c0a66011473021034e7dd4a77740112/bin/tune.py#L185) launches the hyperparameter tuning procedure.
- specifically, it runs [the `objective` function](https://github.com/yandex-research/tabular-dl-tabr/blob/d628ec7e1c0a66011473021034e7dd4a77740112/bin/tune.py#L150) for `n_trials=100` times (see the config).
- the `objective` function, in turn, [calls](https://github.com/yandex-research/tabular-dl-tabr/blob/d628ec7e1c0a66011473021034e7dd4a77740112/bin/tune.py#L153) the `function = "bin.tabr.main"` (see the config).
- the report of each of `n_trials` runs is [saved](https://github.com/yandex-research/tabular-dl-tabr/blob/d628ec7e1c0a66011473021034e7dd4a77740112/bin/tune.py#L157). So you have to ensure that you don't store anything non-trivial in reports, otherwise, it indeed can result in high memory consumption, because they are all stored in memory during the whole tuning process.

**Comment by Yura52** [comment]
Feel free to reopen the issue if needed!
